src/main/scala/intellij/haskell/module/HaskellModuleBuilder.scala

```diff
@@ -218,12 +223,18 @@ object HaskellModuleBuilder {
     }
   }
 
-  private def downloadHaskellPackageSources(project: Project, stackPath: String, haskellPackages: Seq[HaskellPackageInfo], progressIndicator: ProgressIndicator) = {
+  private def downloadHaskellPackageSources(project: Project, libRoot: File, stackPath: String, haskellPackages: Seq[HaskellPackageInfo], progressIndicator: ProgressIndicator) = {
     val step = 0.8 / haskellPackages.size
     var progressFraction = InitialProgressStep
-    haskellPackages.flatMap { packageInfo =>
-      val fullName = packageInfo.name + "-" + packageInfo.version
-      val stdErr = CommandLine.runProgram(Some(project), project.getBasePath + File.separator + LibName, stackPath, Seq("unpack", fullName), 10000, captureOutputToLog = true, logErrorAsInfo = true).map(_.getStderr)
+
+    libRoot.listFiles().filterNot(subDir => haskellPackages.map(_.dirName).contains(subDir.getName)).foreach(FileUtil.delete)
+
+    haskellPackages.filter(packageInfo => {
```

Comment:
Sorry, but I still do not understand this logic.

You have to make distinction between the packages which have to be downloaded and which has to be added to library table. The library table is always cleared before adding library entries.

This is about downloading packages which are not downloaded yet.
```
 haskellPackages.filter(packageInfo => {
      new File(project.getBasePath + File.separator + LibName + File.separator + packageInfo.dirName).exists()
    }) ++ haskellPackages.filter(packageInfo => {
      !new File(project.getBasePath + File.separator + LibName + File.separator + packageInfo.dirName).exists()
    }).
```
is the same as
```
 haskellPackages.filterNot(packageInfo => {
      new File(project.getBasePath + File.separator + LibName + File.separator + packageInfo.dirName).exists()
    }).
```